File: logbook-spring-webflux/src/main/java/org/zalando/logbook/spring/webflux/State.java
package org.zalando.logbook.spring.webflux;

import lombok.RequiredArgsConstructor;

interface State {

    default State with() {
        return this;
    }

    default State without() {
        return this;
    }

    default State buffer(final byte[] message) {
        return this;
    }

    default byte[] getBody() {
        return new byte[0];
    }


    class Buffering implements State {

        private byte[] body;

        @Override
        public State without() {
            return new Ignoring(this);
        }

        @Override
        public State buffer(byte[] message) {
            body = message;
            return this;
        }

        @Override
        public byte[] getBody() {
            return body;
        }
    }

    class Unbuffered implements State {
        @Override
        public State with() {
            return new Offering();
        }
    }

    class Offering implements State {
        @Override
        public State without() {
            return new Unbuffered();
        }

        @Override
        public State buffer(byte[] message) {
            return new Buffering().buffer(message);
        }
    }

    @RequiredArgsConstructor
    class Ignoring implements State {

        private final Buffering buffering;

        @Override
        public State with() {
            return buffering;
        }

        @Override
        public State buffer(byte[] message) {
            return buffering.buffer(message);
        }
    }
}

File: logbook-ktor-common/src/main/kotlin/org/zalando/logbook/common/State.kt
package org.zalando.logbook.common

sealed class State {
    var body: ByteArray = EMPTY_BODY
    open fun with(): State = this
    open fun without(): State = this
    open fun buffer(content: ByteArray): State = this

    class Buffering : State() {
        override fun without(): State = Ignoring(this)
        override fun buffer(content: ByteArray): State = apply { body = content }
    }

    class Unbuffered : State() {
        override fun with(): State = Offering()
    }

    class Offering : State() {
        override fun without(): State = Unbuffered()
        override fun buffer(content: ByteArray): State = Buffering().buffer(content)
    }

    class Ignoring(private val delegate: Buffering) : State() {
        override fun with(): State = delegate
        override fun buffer(content: ByteArray): State = apply { delegate.buffer(content) }
    }
}
